plugins/com.google.cloud.tools.eclipse.appengine.localserver.test/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/ServerPortExtensionTest.java

```diff
@@ -0,0 +1,284 @@
+
+    portExtension.portText.setText("0");
+    assertFalse(portExtension.portDecoration.isVisible());
+  }
+
+  @Test
+  public void testHandlePropertyChanged_nullEvent() {
+    portExtension.handlePropertyChanged(null);
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+  }
+
+  @Test
+  public void testHandlePropertyChanged_nonAppEngineServerType() {
+    PropertyChangeEvent nullPropertyValueEvent = new PropertyChangeEvent(new Object() /* source */,
+        null /* propertyName */, null /* oldValue */, null /* newValue */);
+    portExtension.handlePropertyChanged(nullPropertyValueEvent);
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+    assertFalse(portExtension.portDecoration.isVisible());
+
+    PropertyChangeEvent nonServerTypePropertyEvent =
+        new PropertyChangeEvent(new Object(), null, null, new Object() /* not IServerType */);
+    portExtension.handlePropertyChanged(nonServerTypePropertyEvent);
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+    assertFalse(portExtension.portDecoration.isVisible());
+
+    portExtension.handlePropertyChanged(newNonAppEngineServerTypeEvent());
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+    assertFalse(portExtension.portDecoration.isVisible());
+  }
+
+  @Test
+  public void testHandlePropertyChanged_appEngineServerType() {
+    portExtension.handlePropertyChanged(newAppEngineServerTypeEvent());
+    assertTrue(portExtension.portLabel.isVisible());
+    assertTrue(portExtension.portText.isVisible());
+    assertFalse(portExtension.portDecoration.isVisible());  // Valid port, so not visible.
+  }
+
+  @Test
+  public void testHandlePropertyChanged_showAndHide() {
+    portExtension.handlePropertyChanged(newAppEngineServerTypeEvent());
+    portExtension.handlePropertyChanged(newNonAppEngineServerTypeEvent());
+    assertFalse(portExtension.portLabel.isVisible());
+    assertFalse(portExtension.portText.isVisible());
+    assertFalse(portExtension.portDecoration.isVisible());
+  }
+
+  @Test
+  public void testTriggerPortDecorationOnServerTypeChanges() {
+    portExtension.portText.setVisible(true);
+
+    portExtension.portText.setText("");
+    assertTrue(portExtension.portDecoration.isVisible());
+    assertEquals(portExtension.informationImage, portExtension.portDecoration.getImage());
+
+    portExtension.handlePropertyChanged(newNonAppEngineServerTypeEvent());
+    assertFalse(portExtension.portDecoration.isVisible());
+
+    portExtension.handlePropertyChanged(newAppEngineServerTypeEvent());
+    assertTrue(portExtension.portDecoration.isVisible());
+    assertEquals(portExtension.informationImage, portExtension.portDecoration.getImage());
+
+    portExtension.portText.setText("65536");
```

Comment:
This step is rather testing whether using an invalid value will trigger an error decorator, I think it'd be better in it's own test case (unless it's already covered).